Improved drawing in plot_rag_draw.py example

vighneshbirodkar opened the pull request:
See new figures
![figure_1](https://cloud.githubusercontent.com/assets/3823490/12225426/6c1372dc-b7d8-11e5-8f41-008313359b73.png)
![figure_2](https://cloud.githubusercontent.com/assets/3823490/12225425/6c12ccec-b7d8-11e5-87d8-55f23d0a1039.png)
![figure_3](https://cloud.githubusercontent.com/assets/3823490/12225427/6c16964c-b7d8-11e5-876c-a31bcec9b7d8.png)

vighneshbirodkar commented:
@jni @JDWarner Thoughts ?

jni commented:
@vighneshbirodkar this still could use some tweaking... I am not yet sure how, but these aren't very satisfying...

Here's my thinking: ultimately you want to show both the background image and the (continuous-valued) edge map. This is not possible when using maps like viridis that span the full color range. Instead, we should display the image using a full-range, monochromatic colormap, e.g. greens, and then the edges using a complementary colormap, e.g. something ranging along magenta and orange, that reaches neither black nor white in luminance.

Those were just examples — I don't know what colormaps we should choose, but that's what we should be aiming for imho.

vighneshbirodkar commented:
Or could we dim the image ?

jni commented:
@vighneshbirodkar the problem with dimming the image is that it becomes very difficult to see. The graph by itself is not interesting. What's interesting is the graph superimposed on the image.

stefanv commented:
Or gray + any colormap?
On Jan 12, 2016 09:05, "Juan Nunez-Iglesias" <notifications@github.com>
wrote:

> @vighneshbirodkar <https://github.com/vighneshbirodkar> the problem with
> dimming the image is that it becomes very difficult to see. The graph by
> itself is not interesting. What's interesting is the graph superimposed on
> the image.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-image/scikit-image/pull/1872#issuecomment-170818468>
> .
>

vighneshbirodkar commented:
@stefanv We already support that with the `desaturate` option

JDWarner commented:
Personally, I would be fine with desaturated+color graph overlay as the default and possibly only method.

vighneshbirodkar commented:
Line thickness helps a lot, here are some examples. 

## afmhot
![afmhot](https://cloud.githubusercontent.com/assets/3823490/12279696/b46b83b0-b956-11e5-83bc-ed8ade5bec47.png)

## inferno
![inferno](https://cloud.githubusercontent.com/assets/3823490/12279702/bc2fbcba-b956-11e5-9aa3-4a1a8f9aee58.png)

## jet
![jet](https://cloud.githubusercontent.com/assets/3823490/12279713/d4a8752a-b956-11e5-9f96-acf8432c785b.png)

## magma
![magma](https://cloud.githubusercontent.com/assets/3823490/12279716/da843a10-b956-11e5-9e3a-0aec8b1ac0a9.png)

## plasma 
![plasma](https://cloud.githubusercontent.com/assets/3823490/12279722/e0ee61a0-b956-11e5-8fff-cab3177faa14.png)

## viridis
![viridis](https://cloud.githubusercontent.com/assets/3823490/12279726/e6915e14-b956-11e5-91fd-e0f9f72b3c2a.png)

The script I used is here
https://gist.github.com/vighneshbirodkar/ca5908d513ec7993893a

jni commented:
@vighneshbirodkar @JDWarner @stefanv The problem with the gray cmap is that bright areas and dark areas have low contrast with bright edges / dark edges on most colormaps. (The pixel boundaries don't help, either.)

vighneshbirodkar commented:
@jni can you suggest a background color map and edge color map combination ? I chose to keep the boundaries because they demarcate the region.

jni commented:
@vighneshbirodkar I agree the boundaries are useful. Maybe "mode='subpixel'" would be useful, although you would then have to transform all the other coordinates. Might be worthwhile. Another idea would be to draw the boundaries with some alpha.

For colormaps, obviously have a look [here](http://matplotlib.org/users/colormaps.html) for inspiration. After looking at that for a bit, I think bone (image) with cool (edges) might work quite well, actually! Also gist_heat with winter.

Guidelines for what we should be looking for:

- image itself should span the full luminance range
- graph edges should span a restricted range (avoiding the black and white ends of the spectrum) and use contrasting colors to the image.
- avoid red/green contrast. More colorblind friendly contrasts include blue-orange and green-purple.

Another suggestion would be to use a panel of a few more test images, since we might be overfitting to this coffee cup. =) But, having said that, I think you started with the most difficult example, since the image is very contrasty.

As a complete aside I'd like to give another shout-out to @stefanv and @njsmith for their stellar work on the new MPL colormaps. I keep hunting for their design page to send to various people sending me jet figures, and every time I find it I'm wowed by how thorough it is. Thanks!

vighneshbirodkar commented:
### bone with cool
![bone_cool](https://cloud.githubusercontent.com/assets/3823490/12364346/06f8eb62-bb9c-11e5-9847-fc7c91592ee0.png)


### gist_heat with winter
![gist_heat_winter](https://cloud.githubusercontent.com/assets/3823490/12364353/0b448d84-bb9c-11e5-8a22-0627144faee6.png)

I used an alpha of `.75` for all the edges this time

jni commented:
Hmm, heat/winter doesn't work at all, but bone/cool looks pretty nice to me! What do you think?

vighneshbirodkar commented:
I liked the plasma/viridis/magma over gray plots more.

jni commented:
@vighneshbirodkar I think they are slightly more pleasing to the eye, but we are primarily concerned about a clear representation of the data. Have a look at the top-right corner of the image in those examples: I can't tell the difference between graph edges and fragment boundaries. I think that's a major problem. Similarly, with the top-left and bottom-center, the black of the image becomes muddled with edges and boundaries. It all just looks black. (Plasma is perhaps the least offensive in this regard.) Note that, with SLIC, our brains can complete the grid quite effectively. When we start using more irregular regions, it's going to be more difficult, and clarity of representation will be even more important.

@stefanv @JDWarner any further opinions?

vighneshbirodkar commented:
@stefanv @JDWarner Can you give us your thoughts ? We can make the most popular choice the default.

stefanv commented:
bone + cool looks great to me.  Have you tried bone + magma or bone +
viridis?  I think the above plots were gray + cmap.

JDWarner commented:
@vighneshbirodkar I agree with @stefanv that bone for the underlying image seems to work well. A couple more comparison plots on top of bone and I think we can all be satisfied.

vighneshbirodkar commented:
## bone + afmhot
![afmhot](https://cloud.githubusercontent.com/assets/3823490/12707216/26d88f4e-c85f-11e5-960f-e641b8ae4a7d.png)

## bone + inferno
![inferno](https://cloud.githubusercontent.com/assets/3823490/12707222/2e025764-c85f-11e5-822d-bf87ca72f514.png)

## bone + magma
![magma](https://cloud.githubusercontent.com/assets/3823490/12707226/35076af4-c85f-11e5-9671-c55e85f85c73.png)

## bone + plasma
![plasma](https://cloud.githubusercontent.com/assets/3823490/12707227/3b5173a0-c85f-11e5-9bff-5cc899cb1147.png)

## bone + viridis
![viridis](https://cloud.githubusercontent.com/assets/3823490/12707230/42e5f442-c85f-11e5-8b25-3c31a491483a.png)

vighneshbirodkar commented:
/ping @jni @JDWarner @stefanv 
Could you please take a quick look ?

jni commented:
I like bone + inferno or magma

JDWarner commented:
Bone + inferno for me. Accentuates the difference a little more. 

I could go either way on the naming. I see your point, but `show` reminds one of `plt.show()` from the matplotlib perspective.

jni commented:
I'm happy to keep the name as `draw_rag`. I think there's a way to extract the image from mpl though...?

stefanv commented:
I quite like bone + plasma, as it highlights the grid.  But I am definitely
not going to bike shed this, so please go ahead with whichever option you
prefer!

vighneshbirodkar commented:
@jni We can get the image from mpl but do we will have to do `imshow` and that might interfere with
what the user is already plotting.

jni commented:
@vighneshbirodkar not if you create a new figure with a different backend... ;) Even better might be to figure out whatever mpl is doing to colormap an image... That shouldn't be too difficult... (famous last words)

vighneshbirodkar commented:
@jni 
Does `img_to_colormap` sound like a useful function to you ?

jni commented:
@vighneshbirodkar Not at all! =D I would call it `colormap_image` because I like verbs, though. =)

stefanv commented:
``colormap_image`` ++!

vighneshbirodkar commented:
@jni @stefanv @JDWarner  The colormap PR reached a very anticlimactic end. But nonetheless we have a solution for rendering images with colromaps. However, we are still left with the issue of drawing thick lines.

jni commented:
@vighneshbirodkar what issue is that? I suggest passing a "linewidth" keyword argument that defaults to whatever you think looks good? =)

vighneshbirodkar commented:
@jni The prbolem is we cannot return an image. Unless we use matplotlib to draw and then get the image from the canvas. Although before doing anything I would like to consult @tacaswell

jni commented:
@vighneshbirodkar what do you mean we "cannot return an image"? Are you saying draw_line doesn't support varying line thickness?

At any rate, I think line thickness can wait for a later PR, I would think. However, consulting @tacaswell does seem to be a wise decision at this point. =P

tacaswell commented:
I am super confused what is going on here.  I had seen these images going by and assumed the lines were drawn with mpl `LineCollection` + `scatter` (both of which will take care of the underlying color mapping for you), but I see that you are using skimage to draw the lines.

What is the goal here?  To get nice figures assuming you have mpl, or to get a reasonable png out no matter what?  

If it is the first, I think delegating a bunch of this work to mpl (and taking an `Axes` as an optional input) would make more sense.  By letting Agg do the rasterization you would get better behavior on zooming and by using mpl artist for the edges / centers you could do things like interactively re-map the colors or hook into the `mouse_event` / `picker` framework to add in interactivity (click on a node, get it's numeric weight).  It would also allow you to output the edges as vector graphics.

vighneshbirodkar commented:
@tacaswell If we change the behavior of `draw_rag` to draw with `matplotlib` itself, we can potentially mess up what the user is already trying to draw. Also, the user will need to call `plt.show()` to see the results. 

@jni `draw.line` does not support thickness. The plots with 1 pixel wide lines don't look nearly as good.

vighneshbirodkar commented:
@tacaswell Just to clarify, the plots you see were generated by separate gist, instead of the `draw_rag` function
https://gist.github.com/vighneshbirodkar/ca5908d513ec7993893a

stefanv commented:
How about using anti-aliased lines?
​

jni commented:
@stefanv @vighneshbirodkar I'm inclined to agree with @tacaswell. What else will the user really want to draw on top of this, and why wouldn't they use mpl for it as well? At any rate, since it's in future, we have free reign to fiddle with the API.

vighneshbirodkar commented:
@jni What I mean is if the user is already drawing something we might end up drawing on top of it. Also, I would suggest renaming this to `show_rag`

stefanv commented:
I am happy to use mpl.
​

tacaswell commented:
If you are concerned about messing with other things the user is doing plotting wise I suggest taking `ax` as an optional input and then doing

```python
if ax is None:
    fig, ax = plt.subplots()
```

That way if the user wants to use this in conjunction with other plotting they can pass in an axes to put it on (and if it messes with what ever else is there, that is their problem) or if the user does not pass in an `Axes` makes a new figure which will not (directly) mess with the user (it will do so _indirectly_ if the user is running other code that use `plt.plot` and such to plot to the current axes as the current axes will change via `plt.subplots()`.  If you are paranoid you can set the the current axes back to what it was before

```python
if ax is None:
    old_ax = plt.gca()  # which make create a figure 
    fig, ax = plt.subplots()
    plt.sca(old_ax)
```
but that runs the risk of creating two figures where you only needed one, but such are the risks of using global state).

JDWarner commented:
:+1: for @tacaswell's suggestion. Use mpl across the board; allow a user to the option to pass their own axis to draw on.

vighneshbirodkar commented:
@JDWarner @jni @stefanv 
Please have a look. Here are the new plots. I have removed the `node_color` attribute. Also, I have removed the ability to draw edges using a single color.

### plot_rag_draw.py
![plot_rag_draw_1](https://cloud.githubusercontent.com/assets/3823490/13800722/35e97a28-eb02-11e5-8a14-c5115fd1d91f.png)
![plot_rag_draw_2](https://cloud.githubusercontent.com/assets/3823490/13800723/35f09f10-eb02-11e5-9def-c7d89c8e1930.png)

### plot_rag_boundary.py
![plot_rag_boundary_1](https://cloud.githubusercontent.com/assets/3823490/13800725/46b7c1fc-eb02-11e5-9b4b-a2fa7814880f.png)

For reference, here are their old plots
http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_rag_draw.html#example-segmentation-plot-rag-draw-py
http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_rag_boundary.html

vighneshbirodkar commented:
The colormap magma is not preset in older versions of matplotlib. Should I add it like viridis ?

JDWarner commented:
@vighneshbirodkar IMO, just add it. We probably should ship all four for backwards compatibility in future examples.

jni commented:
I like them! I still think the boundaries are a bit too thick, but that might be inevitable.

vighneshbirodkar commented:
@jni Is that a +1 from you ?

soupault commented:
xref https://github.com/scikit-image/scikit-image/issues/1857

vighneshbirodkar commented:
@jni @stefanv @JDWarner 
Please take a look at the new plots

![plot_rag_boundary_1](https://cloud.githubusercontent.com/assets/3823490/13832517/4c00d28c-ebb3-11e5-9d85-207547523cff.png)
![plot_rag_draw_1](https://cloud.githubusercontent.com/assets/3823490/13832518/4c04e61a-ebb3-11e5-9382-1e7b85023f93.png)
![plot_rag_draw_2](https://cloud.githubusercontent.com/assets/3823490/13832519/4c07ae22-ebb3-11e5-8a82-d241a9f7637f.png)

vighneshbirodkar commented:
I would be glad if someone could review this. Thanks to everyone so far for their patience

jni commented on doc/examples/segmentation/plot_rag_draw.py:
```diff
@@ -8,33 +8,22 @@
-          "mapped between blue and orange.")
-plt.imshow(out)
+fig, ax = plt.subplots()
+ax.set_title('RAG drawn will all edges')
```
will -> with

jni commented on doc/examples/segmentation/plot_rag_draw.py:
```diff
@@ -8,33 +8,22 @@
+fig, ax = plt.subplots()
+ax.set_title('RAG drawn will all edges')
+smap = graph.show_rag(labels, g, img, ax=ax)
+plt.colorbar(smap, fraction=0.03)
```
what is the "fraction" kwarg? Please add a comment to specify what it means.

jni commented on skimage/future/graph/rag.py:
```diff
@@ -413,14 +410,14 @@ def rag_boundary(labels, edge_map, connectivity=2):
-             desaturate=False, in_place=True):
+def show_rag(labels, rag, img, border_color='black', edge_width=1.5,
+             edge_cmap='magma', img_cmap='bone', thresh=np.inf, in_place=True,
+             ax=None):
```
Comment for future reference, you don't need to tackle in this PR, but raise an issue in that case: can borders and edges have alpha? That might make it even more visible.

jni commented on skimage/future/graph/rag.py:
```diff
@@ -431,28 +428,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
-    colormap : colormap, optional
-        Any matplotlib colormap. If specified the edges are colormapped with
-        the specified color map.
+        Color with which the borders of regions are drawn.
```
of -> between would be clearer imho

jni commented on skimage/future/graph/rag.py:
```diff
@@ -431,28 +428,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
+        the image is drawn as it is.
     thresh : float, optional
-        Edges with weight below `thresh` are not drawn, or considered for color
+        Edges with weight above `thresh` are not drawn, or considered for color
```
Hmm, this change makes me wonder... Some RAGs use distance as a metric (in which case this is probably the right thing to do), but others might use similarity, in which case this is the *wrong* thing to do. What do you think about "min_edge_value" and "max_edge_value"?

vighneshbirodkar replied:
I'm not a fan of threshold in the first place. If someone wants to omit some edges they can always remove them from the graph.

And I don't think there is a right or wrong thing to do for any kind of graph, it really depends on what the user is trying to do.

jni replied:
It's totally fine to remove the threshold parameter altogether as far as I'm concerned!

jni commented on skimage/future/graph/rag.py:
```diff
@@ -431,28 +428,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
+        The axes to draw on.
 
-    Returns
+    Reutrns
```
LOL Worst. Change. Ever. =P

vighneshbirodkar replied:
I don't know how I managed to do this but I laughed for the better part of a minute reading your comment.

jni commented on skimage/future/graph/rag.py:
```diff
@@ -431,28 +428,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
-    out : ndarray, shape (M, N, 3)
-        The image with the RAG drawn.
+    smap : matplotlib ScalarMappable
+         A ScalarMappable for the edges that can be used to show colorbars.
```
I have no idea what a ScalarMappable is, so perhaps adding a hyperlink reference to this would be useful.

jni replied:
Actually I just saw the example below. Maybe say, "that can be passed to the `pyplot.colorbar()` to show the scale of the colormap used.

tacaswell replied:
There is magic you can use with Sphinx that will auto-link to the mpl docs (`intersphinx_mapping` is the key in  `confg.py`)

jni commented on skimage/future/graph/rag.py:
```diff
@@ -461,20 +457,32 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
-        img = color.gray2rgb(img)
-
+    if ax is None:
+        fix, ax = plt.subplots()
```
fix -> fig. ie you should fix `fix`. =P

tacaswell commented on skimage/future/graph/rag.py:
```diff
@@ -491,31 +499,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
+            r1, c1 = map(int, rag.node[n1]['centroid'])
+            r2, c2 = map(int, rag.node[n2]['centroid'])
+            col = smap.to_rgba([data['weight']])[0][:-1]
+            ax.add_artist(plt.Line2D([c1, c2], [r1, r2],  color=col,
```
I would use a `LineCollection` for this, you get better prefromance and mpl will take care of the norm/color mapping for you.

jni commented on skimage/future/graph/rag.py:
```diff
@@ -461,20 +457,32 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
+    if img_cmap is None:
+        if img.ndim < 3:
+            raise ValueError('The image needs to be RGB')
+        if img.ndim < 4:
```
What's all this? It seems brittle. If I give you a 3D grayscale image, you'll give me the first three columns blended as RGB and ignore the rest.

On the other hand, while we're assuming things, why not assume grayscale when the image is 2D and they've used colormap=None? Or, at least, perhaps the ValueError should specify that it should be RGB *or* a colormap should be selected.

jni commented on skimage/future/graph/rag.py:
```diff
@@ -461,20 +457,32 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
+        out = color.rgb2gray(img)
+        out = img_cmap(out)
+
+    edge_cmap = cm.get_cmap(edge_cmap)
```
I think we've discussed this, but can you remind me again, why does this work when `edge_cmap` is `'viridis'` and the matplotlib version is < 1.5?

jni replied:
Nevermind, just found it below, thank you! =)

jni commented:
@vighneshbirodkar I've made some minor comments. Other than those, this is good, I'll be happy to (finally) merge! Thank you!

vighneshbirodkar commented on skimage/future/graph/rag.py:
```diff
@@ -461,20 +456,30 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
 
-    edge_color = cc.to_rgb(edge_color)
-    node_color = cc.to_rgb(node_color)
+    if img_cmap is None:
```
@jni This seems ok to you ?

jni replied:
@vighneshbirodkar I would additionally check that `img.shape[2] <= 4`

vighneshbirodkar replied:
@jni Why not exactly equal to 3 ?

vighneshbirodkar replied:
I would do `3 <= img.shape[2] <= 4`

soupault replied:
`img.shape[2] in [3, 4]`?

jni replied:
I believe @soupault's version to be both the fastest, most concise, and most standard (ie that pattern exists elsewhere in skimage)

vighneshbirodkar commented:
The 2nd plot in `plot_rag_draw.py` looks like this
![plot_rag_draw_2](https://cloud.githubusercontent.com/assets/3823490/13914518/382339ba-ef25-11e5-962a-158884014aec.png)

vighneshbirodkar commented:
@jni @soupault If you can have another look. It shouldn't be far from merge now

jni commented on skimage/future/graph/rag.py:
```diff
@@ -506,6 +506,7 @@ def show_rag(labels, rag, img, border_color='black', edge_width=1.5,
     # Defining the end points of the edges
     # The tuple[::-1] syntax reverses a tuple as matplotlib uses (x,y)
     # convention while skimage uses (row, column)
+    ll_ = [(rag.node[n1]['centroid'], rag.node[n2]['centroid']) for (n1, n2) in rag.edges_iter()]
```
This goes unused?

soupault replied:
@jni https://github.com/scikit-image/scikit-image/pull/1872/files#diff-ad47b78e529d8acd83e9c6bbb84874bbR513

soupault commented on skimage/future/graph/rag.py:
```diff
@@ -429,30 +426,28 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
         If set, the RAG is modified in place. For each node `n` the function
         will set a new attribute ``rag.node[n]['centroid']``.
+    ax : matplotlib axes, optional
+        The axes to draw on.
```
I'm reviewing this for the first time and this confused me.
Probably, as far as `ax` is `optonal`, it's worth to add a note on default behaviour.

vighneshbirodkar commented on skimage/future/graph/rag.py:
```diff
@@ -489,33 +494,25 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
 
-        circle = draw.circle(r1, c1, 2)
-        out[circle] = node_color
+    lc = LineCollection(lines, linewidths=edge_width, cmap=edge_cmap)
```
@jni. LineCollection is used here

jni commented:
@vighneshbirodkar and `ll_`?

vighneshbirodkar commented:
@jni Fixed that

jni commented on skimage/future/graph/rag.py:
```diff
@@ -429,30 +426,28 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
     rag : RAG
         The Region Adjacency Graph.
     img : ndarray, shape (M, N, 3)
```
shape (M, N[, 3]), right?

jni commented on skimage/future/graph/rag.py:
```diff
@@ -429,30 +426,28 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
-        The image with the RAG drawn.
+    lc : :py:class:`matplotlib.collections.LineCollection`
+         A colection of lines that represent the edges of the graph. It can be
+         passed to the :meth:`matplotlib.figure.Figure.colorbar` function.
```
:100:

jni commented on skimage/util/colormap.py:
```diff
@@ -258,4 +259,266 @@
+
+
 viridis = LinearSegmentedColormap.from_list('viridis', viridis_data)
+cm.register_cmap('viridis', viridis)
```
I would do this only if they are not already registered, so as not to overwrite whatever ships with matplotlib >1.5, which might have some improvements over what we ship here, for example.

jni commented:
@vighneshbirodkar this is almost done! Just two very minor comments.

vighneshbirodkar commented:
@jni @soupault All tests pass now. Please have look

vighneshbirodkar commented:
Looks like the AppVeyor build timed out!

jni commented:
@vighneshbirodkar it's been doing that. No big deal. =)

This gets a 👍 from me! Thanks for your infinite patience, @vighneshbirodkar! =)

soupault commented:
@vighneshbirodkar have you tried to supplement these new plots with RAG vertices? Can't see any point in the discussion where we decided to drop them.
The PR looks good to me, but I'd like to clarify this first.

jni commented:
@soupault I would say that vertices never really added much to the drawing, and we spent quite enough time bikeshedding on this...

vighneshbirodkar commented:
@soupault

I mentioned that I removed the node_color attribute.@JDWarner once pointed that out and even I feel that node color and is one extra parameter the user has to choose while it does nothing to improve the drawing. That's also the reason I removed the ability to draw all edges with a single color.

And I feel this function should just get out of the user's way. It's not doing any work as such and it's only purpose is to provide insight on what to do next. The less time the user spends customising this, the better.

soupault commented:
@jni @vighneshbirodkar thanks for your answers! Cannot see any obstacles to finally get this alive then.
Thank you for your patience, @vighneshbirodkar.
Merging.

stefanv commented:
\o/

soupault commented:
Everything regarding this PR is fine:
http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_rag_boundary.html
http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_rag_draw.html

But, Google has [this page indexed](http://scikit-image.org/docs/dev/auto_examples/plot_rag_draw.html), which is actually a mess: the plotted images do not correspond to the code. Can we do anything about that? /cc @emmanuelle

vighneshbirodkar commented:
@soupault Can't we just delete that page ?

soupault commented:
Only manually, I suppose. I don't have these permissions.
Btw, not just this page should be removed, but also the pages for another examples ([1](http://scikit-image.org/docs/dev/auto_examples/plot_regional_maxima.html), [2](http://scikit-image.org/docs/dev/auto_examples/plot_equalize.html), etc). This should have been accounted in https://github.com/scikit-image/scikit-image/pull/1823 :/.

stefanv commented:
Our doc upload script should be updated to not only update and upload new files but to also remove old files.  @soupault I'll give you permissions on that repo as well.
